src/Controllers.tsx

```diff
@@ -111,20 +80,33 @@ export function Controllers({ rayMaterial = {}, hideRaysOnBlur = false }: Contro
       ),
     [JSON.stringify(rayMaterial)] // eslint-disable-line react-hooks/exhaustive-deps
   )
-  React.useMemo(() => extend({ ControllerModel }), [])
-
-  // Send fake connected event (no-op) so models start loading
-  useIsomorphicLayoutEffect(() => {
-    for (const target of controllers) {
-      target.controller.dispatchEvent({ type: 'connected', data: target.inputSource, fake: true })
+  React.useMemo(() => extend({ XRControllerModel }), [])
+
+  const handleControllerModel = useCallback((xrControllerModel: XRControllerModel | null, target: XRController) => {
+    if (xrControllerModel) {
+      target.xrControllerModel = xrControllerModel
+      if (target.inputSource?.hand) {
+        return
+      }
+      if (target.inputSource) {
+        modelFactory.initializeControllerModel(xrControllerModel, target.inputSource)
+      } else {
+        console.warn('no input source on XRController when handleControllerModel')
+      }
+    } else {
+      if (target.inputSource?.hand) {
+        return
+      }
+      target.xrControllerModel?.disconnect()
+      target.xrControllerModel = null
     }
-  }, [controllers])
+  }, [])
 
   return (
     <>
       {controllers.map((target, i) => (
         <React.Fragment key={i}>
-          {createPortal(<controllerModel args={[target]} />, target.grip)}
+          {createPortal(<xRControllerModel ref={(r) => handleControllerModel(r, target)} args={[]} />, target.grip)}
```

Comment:
Need to move `target` to `useCallback` deps so we can get a stable reference here between successive renders.

Reply:
Ooh, that's a good catch, thank you. I wrote a test and fixed it by moving it to ControllerModel component that will accept target now in 3a5039420f19b66ea1b1074ad88d9eb11739beb5